### ISSUE TITLE
Unsuccessful client disconnect

### ISSUE THREAD

**Issue by HollowTooth** [gammazero/nexus#99]
Hi!

I'm having a bit of an issue with the Nexus router and/or client.
I regularly program my code on a laptop, and whenever I accidentaly leave a client (running on the laptop) online, and close the laptop (hibernate), whenever I reopen the laptop, or even if I don't and try to connect that same client (with the same RPC functions), the router returns a procedure_already_exists error. It wouldn't be sooo much an issue, if it did that only once, and then found out, that the client is no longer connected, but it does that for ... about an half an hour or more, then suddenly it realizes, the client is no longer online. Is there anyway to change the way the router behaves (perhaps a timeout, or a check, whenever router returns procedure_already_exist, to check the client of the procedure).

At this point and time, I don't know what causes the router to magically remove the node and the RPC functions, but a controllable way to do this would be really nice, as it usually causes me to restart the router, which in some cases, makes me restart every other node (which can take sometime).

It's not a critical issue, but I hope you can help me on this one :)
( In the case, that there is currently no solution, I could try to fix this in a fork)

**Comment by gammazero** [comment]
Normally when the client session dies, due to socket being closed, the router will clean up the client's session, which will also clean up any subscriptions and registrations.  I will look into this right away to see if this is not behaving correctly.  Thank you for reporting it.

**Comment by gammazero** [comment]
When the client running on your laptop goes offline, it does not close the websocket (or TCP socket) connection; a FIN packet is not sent to the server.  The server does not know the client is gone, and maintains the session and all registrations and subscriptions.  Eventually, the TCP keep-alive on the websocket fires, retries, retries, and eventually the server drops he connection when it determines it is dead.

There are a couple of ways for me to fix this:
1. Use a websocket ping-pong to check for responsiveness of the websocket peer.
2. Allow server keep-alive to be configurable, so that the dead client connection can be detected and closed sooner.

I chose 2, since the the ping-pong would require an extra goroutine per client or to wake up periodically from the websocket receive, both things I would prefer to avoid.  Using TCP keep-alive works well to accomplish and requires minimal new code.  See PR #100.

**Comment by gammazero** [comment]
A way to test this is to connect your client, running on your laptop, to a nexus router running somewhere else.  Then turn off wifi or unplug laptop network and kill the client. Re-enable network and see how long it takes for server to detect dead client.  Then adjust the server keep-alive time to your preferences and restart the server.

**Comment by gammazero** [comment]
I merged PR #100.  Set the websocket server keep-alive like this:
```go
wsServer := router.NewWebsocketServer(nxsRouter)
// Set keep-alive period to 30 seconds.
wsServer.KeepAlive = 30 * time.Second
// Run websocket server.
wsCloser, err := wsServer.ListenAndServe(wsAddr)
```
If this works for you, please close issue.

**Comment by HollowTooth** [comment]
First of all, thanks for responding soo quickly!
Secondly, I did some tests today with PR #100 .
The client DID successfully reconnect, but only after 3~4 minutes (I set the `.KeepAlive` to 10 seconds).
I'm guessing that the keep alive has a certain amount of tries before it defines the session as closed ?

I'll just use put the keep alive to a second, and hope it can reconnect as fast as possible.
Overall it works a lot better now, and is manageable. Closing issue.
